hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/sink/prometheus/PushGatewaySink.java

```diff
@@ -0,0 +1,191 @@
+
+            //ignore specific tag which includes sub-hierarchy
+            if (NUM_OPEN_CONNECTION_SPERUSER.equals(tagName)) {
+              continue;
+            }
+            labelNames[index] = tagName;
+            labelValues[index] =
+                tag.value() == null ? NULL : tag.value();
+            index++;
+          }
+
+          switch (metrics.type()) {
+          case GAUGE:
+            Gauge.build(key, key)
+                .labelNames(labelNames)
+                .register(registry)
+                .labels(labelValues)
+                .set(metrics.value().doubleValue());
+            break;
+          case COUNTER:
+            Counter.build(key, key)
+                .labelNames(labelNames)
+                .register(registry)
+                .labels(labelValues)
+                .inc(metrics.value().doubleValue());
+            break;
+          default:
+            break;
+          }
+          LOG.debug(
+              "register succeed, metrics name is: " + key + " Type is :" + metrics.type().toString()
+                  + " tag is: " + Arrays.toString(labelNames)
+                  + " tagValue is " + Arrays.toString(labelValues)
+                  + " Value is : " + metrics.value().toString());
+        }
+      }
+      pg.push(registry, jobName, groupingKey);
+      LOG.info("pushing succeed");
+    } catch (Exception e) {
+      LOG.error("pushing job's metrics to gateway is failed ", e);
+    }
+  }
+
+  /**
+   * Convert CamelCase based names to lower-case names where the separator is the underscore, to
+   * follow prometheus naming conventions.
+   */
+  public String getMetricsName(String recordName,
+      String metricName) {
+    String baseName = StringUtils.capitalize(recordName)
+        + StringUtils.capitalize(metricName);
+    baseName = baseName.replace('-', '_');
+    String[] parts = SPLIT_PATTERN.split(baseName);
+    return String.join("_", parts).toLowerCase();
+  }
+
+  Map<String, String> parseGroupingKey(final String groupingKeyConfig) {
+    if (!groupingKeyConfig.isEmpty()) {
+      Map<String, String> groupingKey = new HashMap<>();
+      String[] kvs = groupingKeyConfig.split(";");
+      for (String kv : kvs) {
+        int idx = kv.indexOf("=");
+        if (idx < 0) {
+          LOG.warn("Invalid prometheusPushGateway groupingKey:" + kv + ", will be ignored");
+          continue;
+        }
+
+        String labelKey = kv.substring(0, idx);
+        String labelValue = kv.substring(idx + 1);
+        if (StringUtils.isEmpty(labelKey) || StringUtils.isEmpty(labelValue)) {
+          LOG.warn("Invalid groupingKey labelKey:" + labelKey + ", labelValue:" + labelValue
+              + " must not be empty");
+          continue;
+        }
+        groupingKey.put(labelKey, labelValue);
+      }
+      return groupingKey;
+    }
+    return Collections.emptyMap();
+  }
+
+  @Override
+  public void flush() {
+
+  }
+
+  @Override
+  public void close() throws IOException {
```

Comment:
shouldn't you shut down push gateway here?

Reply:
> push gateway

Push gateway is not a port service and cannot be closed.